### ISSUE TITLE
GPS location is wrong, but is correct i  other app

### ISSUE THREAD

**Issue by dbf256** [organicmaps/organicmaps#6739]
I'm on train. Locus Maps show me exactly on the rail, while OM location jumps, displayed with circle and is far away from valid point. 
I saw this when I'm on pedestrian path. For me it looks like issues with location started some time ago and it was better earlier. 
I have Xiaomi with google services enabled in OM.
![Screenshot_2023-11-30-20-03-06-586_menion android locus pro](https://github.com/organicmaps/organicmaps/assets/2550295/bffb13ab-b745-457d-b25f-6da3d527572c)

![Screenshot_2023-11-30-20-03-19-627_app organicmaps](https://github.com/organicmaps/organicmaps/assets/2550295/bac4e91d-cf99-4cf7-b083-45778b6ea088)
![Screenshot_2023-11-30-20-03-37-896_app organicmaps](https://github.com/organicmaps/organicmaps/assets/2550295/a83017c8-a605-4747-ba9e-ac708b714f90)

**Comment by biodranik** [comment]
It looks like a bug with Google Play services. It provides rough network-based location (large circle radius) instead of GPS, and even doesn't provide an actual position (grey arrow means an outdated position).

Try to disable Google Services in settings, or update them on your device.

**Comment by biodranik** [comment]
Yes, you can enable logs in OM settings and then share them with yourself via "Report a bug".

**Comment by dbf256** [comment]
@biodranik  It still bad when Locus is fine even when google services is disabled.
Locus has precise location and OM has large circle and center of the circle is not in valid point.

![Screenshot_2023-12-19-13-17-12-332_menion android locus pro](https://github.com/organicmaps/organicmaps/assets/2550295/331f7336-60e4-4ca5-9e55-cb8a7298749d)

![Screenshot_2023-12-19-13-16-57-268_app organicmaps](https://github.com/organicmaps/organicmaps/assets/2550295/52ec554b-4bfd-44bc-91ca-579abd912599)

![Screenshot_2023-12-19-13-20-05-035_app organicmaps](https://github.com/organicmaps/organicmaps/assets/2550295/129f088f-14e5-42d8-8159-819cafa720fe)

**Comment by biodranik** [comment]
@dbf256 please check if you provided Fine (precise) location permissions to Organic Maps. Looks like you allowed only a rough location permissions for OM.

**Comment by dbf256** [comment]
> @dbf256 please check if you provided Fine (precise) location permissions to Organic Maps. Looks like you allowed only approximate location permissions for OM.

I see only simple Location permission in my phone and it is granted to OM and Locus.
![Screenshot_2023-12-19-22-59-06-288_com google android permissioncontroller](https://github.com/organicmaps/organicmaps/assets/2550295/a3df149d-cda0-48a5-b467-830e3e1136a0)
![Screenshot_2023-12-19-22-58-04-373_com google android permissioncontroller](https://github.com/organicmaps/organicmaps/assets/2550295/1b74ea9c-2a00-42d4-a006-2c387e109368)
![Screenshot_2023-12-19-22-54-40-092_com google android permissioncontroller](https://github.com/organicmaps/organicmaps/assets/2550295/9910c4ed-5400-4456-b7bb-31f04c08c6bc)

**Comment by biodranik** [comment]
@dbf256 what is your Android version?

**Comment by dbf256** [comment]
> @dbf256 what is your Android version?
> 
> Can you please install the latest OM version, enable logs in OM settings, go outside, start detecting your location, wait for a few minutes, and then send us logs using "Report a bug"?

![Screenshot_2023-12-20-10-36-07-253_com android settings~01](https://github.com/organicmaps/organicmaps/assets/2550295/0c752312-22d9-46eb-b8db-4976c9638658)

Here are the details about OS. I'll try to record logs

**Comment by dbf256** [comment]
@biodranik logs are send using Report Issue button. Had the same issues today also - so happy to do testing/send more logs if required. The issue is quite annoying - I was not able to conveniently use OM for navigation today and switched for LM.

**Comment by biodranik** [comment]
@dbf256 is your system time set manually? If yes, can you please try to set it to automatic?

**Comment by dbf256** [comment]
> @dbf256 is your system time set manually? If yes, can you please try to set it to automatic?

Yes, it was set to manual, I switched it to Network Time and will do some testing.

**Comment by dbf256** [comment]
@biodranik  thanks for your help, gps works fine after time changed. Not sure is it a common issue, but maybe display a warning if it is critical.

**Comment by biodranik** [comment]
@rtsisyk @vng we can detect wrong time by comparing the current system time with the GPS time and warn the user that his device's clock is too off.
Or we can store the time when GPS was received, and compare it with the provided network location time, then comparison should work properly.
Or we can completely turn off the network location to save the battery if GPS is available. And as the next improvement, turn on the network location again if GPS was not available for too long.

The failed case that should work properly after the fix:
```
12-21 10:43:31.011 (main) D/app.organicmaps.location: LocationHelper.java:191 onLocationChanged(): The new Location[gps 50,085066,14,440610 hAcc=4 et=+8d2h4m15s28ms etAcc=+1ms alt=272.150146484375 vel=1.9 bear=242.7 vAcc=3 sAcc=0 bAcc=2 {Bundle[mParcelledData.dataSize=96]}] is worse than the last Location[network 50,085462,14,440217 hAcc=24 et=+8d2h6m15s970ms alt=257.8999938964844 vAcc=2 sAcc=??? bAcc=??? {Bundle[mParcelledData.dataSize=332]}]
```

The current code:
```Java
public static boolean isLocationBetterThanLast(@NonNull Location newLocation, @NonNull Location lastLocation)
  {
    // As described in isAccuracySatisfied, GPS may have zero accuracy "for some reasons".
    if (isFromGpsProvider(lastLocation) && lastLocation.getAccuracy() == 0.0f)
      return true;

    double speed = Math.max(DEFAULT_SPEED_MPS, (newLocation.getSpeed() + lastLocation.getSpeed()) / 2.0);
    double lastAccuracy = lastLocation.getAccuracy() + speed * LocationUtils.getTimeDiff(lastLocation, newLocation);
    return newLocation.getAccuracy() < lastAccuracy;
  }
```
